feat: add era pre-(fetching/processing)

KolbyML opened the pull request:
https://github.com/ethereum/trin/pull/1356 PR should be reviewed first

### What was wrong?
Era1 files would block executing as we had to wait for them to download then recover the senders
### How was it fixed?
download and process the next era1 file in the background while executing

morph-dev commented on trin-execution/src/era_manager.rs:
```diff
@@ -23,37 +47,86 @@ impl EraManager {
-            .find(|file| file.contains(&format!("mainnet-{epoch_index:05}-")))
-            .expect("to be able to find era1 file");
+            .find(|file| file.contains(&format!("mainnet-{:05}-", epoch_index + 1)))
+            .expect("to be able to find era1 file")
```
What if `epoch_index` corresponds to the last epoch? This would fail, and so would `get_block_by_number`, which isn't expected behavior?.

morph-dev commented on trin-execution/src/era_manager.rs:
```diff
@@ -23,37 +47,86 @@ impl EraManager {
+    ) -> anyhow::Result<Vec<BlockWithRecoveredSenders>> {
+        info!("Fetching the next era file");
+        let start = Instant::now();
+        while self.current_era.is_some() && self.next_era.lock().is_none() {
```
I would say that this is dangerous approach for dealing with async tasks.

For example, what if task spawn at line 109 fails (for example, network issue)? It will just finish without setting `next_era`, and this one would fail after waiting for 20 minutes.

morph-dev reviewed (changes requested):
I like the optimization idea, but I think there are few cases that are not handled correctly in regard to async tasks and usage of the EraManager (see line comments).

Here is the approach that I would take.
Instead of having `next_era: Arc<Mutex<Option<...>>>`, I would have it like this:
`next_era: Option<JoinHandle<ProcessedEra>>` (where `JoinHandle` is the object returns by `tokio::spawn`).

Also, considering that maybe not all users of `EraManager` plan to process data sequentially, I would have a parameter that would enable/disable this optimization.

Happy to have a call if you want to discuss this into more details.

KolbyML commented on trin-execution/src/era_manager.rs:
```diff
@@ -1,15 +1,39 @@
 
 pub struct EraManager {
-    current_era: Option<Era1>,
+    current_era: Option<Vec<BlockWithRecoveredSenders>>,
```
![image](https://github.com/user-attachments/assets/e686d80a-31e5-459c-9b08-fac5e5dcf3cb)

This is my issue with premature review's like this

KolbyML commented:
Me and Milos had a call. I am going to close this PR and open a new one once I've implemented Era support. I already implemented some of the feedback, and others I think is very useful. I made this PR a little pre-mature as I supporting Era requires almost fully rewriting it to an extent
